SimEngine.py

```diff
@@ -96,16 +99,28 @@ def runToCompletion(self,startFunc):
         
         return self._currentTime
     
-    def schedule(self,ts,cb,selfDestruct=False):
+    def schedule(self,ts,cb,tag='None'):
         # add new event
-        self.events += [(ts,cb,selfDestruct)]
+        self.events += [(ts,cb,tag)]
         # reorder list
         self.events  = sorted(self.events, key = lambda e: e[0])
         # release semaphore (increments its internal counter)
         self.semNumEvents.release()
-    
+
+        # reorder list
+        self.events  = sorted(self.events, key = lambda e: e[0])
+
+        # release semaphore (increments its internal counter)
+        self.semNumEvents.release()
+
+    def cancelEvent(self, tag):
+        for (idx,event) in enumerate(self.events):
```

Comment:
this iteration over the list doesn't work, because you are popping elements dynamically.

Assuming you `l` is your list and `r` is what you want to remove:
```
l = ['a','b','c','c','d']
r = 'c'
```

this is *wrong*, it results in `['a', 'b', 'c', 'd']`
```
for (idx,e) in enumerate(l):
    if e == r:
        l.pop(idx)
```

this is *right*, it results in `['a', 'b', 'd']`
```
idx = 0
while idx<len(l):
    if l[idx]==r:
        l.pop(idx)
    else:
        idx += 1
```